crates/amaru/src/bin/amaru/pid.rs

```
@@ -0,0 +1,86 @@
+// Copyright 2025 PRAGMA
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+use std::{
+    fmt::Display,
+    fs,
+    path::{Path, PathBuf},
+    process::{self, Command},
+};
+
+pub struct PIDFile {
+    path: PathBuf,
+    pid: u32,
+}
+
+impl PIDFile {
+    pub fn new<P: AsRef<Path>>(path: P) -> Result<Self, Box<dyn std::error::Error>> {
+        let path = path.as_ref().to_path_buf();
+        let pid = process::id();
+
+        if let Some(parent) = path.parent() {
+            fs::create_dir_all(parent)?;
+        }
+
+        if path.exists()
+            && let Ok(existing_pid) = fs::read_to_string(&path)
+            && let Ok(existing_pid) = existing_pid.trim().parse::<u32>()
+            && process_exists(existing_pid)
+        {
```

That is very strange. The process id is unique to our process; so even if _Amaru_ is already running, there's no chance that another instance of Amaru will have the same process id. 

All-in-all, I am not sure what this is trying to prevent?

---

My thought is that this is to prevent clobbering another processes PID file;

For example, consider:
- You start Amaru with PID path `a/b/c.pid`
- Amaru crashes without a chance to clean up the PID file, and gets automatically restarted
- It starts back up, sees that the file exists (ok...), parses it (ok...), but the PID inside the file isn't running, so we're ok to overwrite this

On the other hand, consider:
- You start Amaru with PID path `a/b/c.pid`
- You accidentally start amaru again with the same command line arguments
- It starts up, sees that the file exists (ok...), parses it (ok...), but the PID inside the file *is* running (it's the previous copy of amaru), so we need to report an error to prevent clobbering that PID file

Finally, consider:
- You start {some other process} with PID path `a/b/c.pid`
- You start Amaru with PID path `a/b/c.pid`
- It starts up, sees that the file exists (ok...), parses it (ok...), but the PID inside the file *is* running (it's {some other process}), so we need to report an error to prevent clobbering that other processes PID file